FCCEUD-61: Document new support and requirements for xrefs in Pv2 (#613)
redhataccess/pantheon: pantheon-bundle/src/main/resources/SLING-INF/content/docs/modules/con_guidance-on-using-xrefs.adoc

[id="guidance-on-using-xrefs_{context}"]
= Guidance on using xrefs

[role="_abstract"]

> To ensure that internal cross-references, xrefs, work when you publish your content, use only inter-document xrefs. Although AsciiDoc supports different forms of xref, {ProductShortName} validates only inter-document xrefs. If validation fails, {ProductShortName} flags invalid xrefs.

Author: you can use xref:some-id[], but Pantheon won't check if it works.

> To ensure that internal cross-references, xrefs, work when you publish your content, use only inter-document xrefs. Although AsciiDoc supports different forms of xref, {ProductShortName} validates only inter-document xrefs. If validation fails, {ProductShortName} flags invalid xrefs.

Author: yep, see https://docs.asciidoctor.org/asciidoc/latest/macros/inter-document-xref/ specifically: "These problems can be alleviated by using an inter-document xref:"

NOTE: {ProductShortName} only validates xrefs within a repository.

The following are the syntaxes of inter-document xref:

* `\xref:document-b.adoc#target-section[Target section]`

> * `\xref:../modules/document-c.adoc[Content C]`

Reviewer: Who is responsible for ensuring that `../modules` directory will exist for the lifetime of the title? (ie is there a contract for url structure as delivered by pantheon?

Author: Interesting point, I thought that it did not matter where the file was. @xdavidson is there a requirement that the repos follow a specific structure for xrefs to work?  

For example, if the module `document-a` was in `topics/xyz/` and module `document-b`, was in `topics/abc/` would an xref in `document-b` - `xref:../xyz/document-a.adoc[Content A]` work?

Thanks Paul, I'll update the text based on the answer!

Reviewer 2: @theashiot @pwright Using xrefs is not dependent on any particular directory structure or naming conventions.

Paul, if you have a file in a directory name `modules` and then at some point change that directory name from `modules` to something else, such as `smurfs`, then you will need to:
* reimport/update your documentation repo in Pv2
* change any xrefs to target `smurfs` instead of `modules`
These are not specific to xrefs.
Did I answer your question @pwright?

Reviewer: @stoobie yes, I think so... my understanding of your answer is that Pv2 deploys the directory structure of your repo and therefore you can create xrefs that work in plain old asciidoc- Pv2 will allow those resolve correctly.

Reviewer 2: @pwright Yes. Pv2 imports your Git repo as-is.

> [dicsrete]

Reviewer 2: s/dicsrete/discrete

alternatively, get rid of the h2 and use .Headings instead.

Author: fixed the spelling of discrete

== Example of using xrefs

In this example, the following are the source and targets:

source document:: 
* A file `document-a.adoc`.

Targets:: 

* A section called _target section_ in the file _document-b.adoc_.
* A file _document-c.adoc_.

Use xrefs as follows:

* To cross-reference section target section in document-b.adoc from document-a.adoc, use the following form:

> ** `\xref:document-b.adoc#target-section[target section]`

Reviewer 2: ```suggestion
+
`\xref:document-b.adoc#target-section[Target section]`
```

+

> where `target-section` is the anchor for _target section_.

Reviewer 2: ```suggestion
where `target-section` is the anchor for _Target section_.
```

* To cross-reference document document-c.adoc from document-a.adoc, use the following form:

** `\xref:../modules/document-c.adoc[Section C]`
+
where `../modules/document-c.adoc` is the path to _document-c.adoc_ relative to _document-a.adoc_.

> IMPORTANT: Do not use xrefs with empty `[]`.

Reviewer 2: "...with empty brackets"

alternatively: Always include text within the square brackets in an xref.

Author: changed to - "Always include text within the square brackets in an xref."

> IMPORTANT: Always include text within the square brackets in an xref.

Reviewer 2: ```suggestion
[IMPORTANT]
====
Always include text within the square brackets in an xref.
====
```
Is this a Pantheon requirement?

Author: yes, it's documented in the pv2 migration checklist

[discrete]
== xfref validation in {ProductShortName}

{ProductShortName} checks xrefs when you:

> * import a repository from git to pantheon,

Reviewer 2: ```suggestion
* import a Git repository to {ProductShortName},
```

Author: done!

* regenerate draft HTML for your imported content,
* publish your content.

> If {ProductShortName} finds any xref-errors, it displays a red alert icon next to the module or assembly with incorrect xref on the search page. Expand the module or assembly with errors to find the xrefs that are incorrect. Use this information to correct the incorrect xrefs in the source and re-import your content.

Reviewer 2: ```suggestion
If {ProductShortName} finds any errors from xrefs, it displays a red alert icon next to the module or assembly with the source of the error on the search page. Expand the module or assembly to find the broken xrefs. After you fix the xrefs in the Asciidoc source, re-import your content.
```

> NOTE: The order in which modules were imported can result in false validation errors. Regenerate draft HTML for the flagged content and the target content to check this.   

Reviewer 2: Make this an Asciidoc note.
